### DISCUSSION TITLE
exceed memory limit

### DISCUSSION THREAD

**Discussion by qtbui159** [libvips/libvips#3740] [Q&A] [answered]
I have a 32-bit program use this code. 

```
VipsImage* big = vips_image_new_from_file("C:\\Users\\Spike\\Desktop\\2-12.jpg", nullptr); // 30000x20000 jpg
VipsImage* small = vips_image_new_from_file("C:\\Users\\Spike\\Desktop\\ffsc.jpg", nullptr); // 1080x2400 jpg
int ret = vips_draw_image(big, small, 0, 0, nullptr); // ret = -1
string errorMessage = vips_error_buffer(); // 'vips_tracked: out of memory --- size == 2414MB
```

How can i make it run correctly?

**Comment by jcupitt** [comment, marked as answer]
Hello @qtbui159,

The `draw_*()` functions are there for paintbox-style programs which need to make small, in-place, iterative modifications to an image, so they need to keep the whole thing in memory. They can't do large images, and they are not threaded.

https://www.libvips.org/API/current/libvips-draw.html#libvips-draw.description

To paste one image into another, use `insert`, eg.:

```C
VipsImage* big = vips_image_new_from_file("C:\\Users\\Spike\\Desktop\\2-12.jpg", nullptr); // 30000x20000 jpg
VipsImage* small = vips_image_new_from_file("C:\\Users\\Spike\\Desktop\\ffsc.jpg", nullptr); // 1080x2400 jpg
VipsImage *out;
if (vips_insert(big, small, &out, 0, 0, nullptr)) { ... error handling ... }
```

It'll run in very little memory, and it'll use all your CPU cores.

**Reply by qtbui159** [reply]
thx for your fast replying @jcupitt .
vips_insert solve my problem.
